src/passes/Asyncify.cpp

```diff
@@ -1214,6 +1244,234 @@ struct AsyncifyFlow : public Pass {
+    public:
+      CountersBuilder(Module& wasm) : Builder(wasm) {}
+      Expression* makeInc(int amount = 1) {
+        return makeGlobalSet(
+          ASYNCIFY_CATCH_COUNTER,
+          makeBinary(AddInt32,
+                     makeGlobalGet(ASYNCIFY_CATCH_COUNTER, Type::i32),
+                     makeConst(int32_t(amount))));
+      }
+      Expression* makeDec(int amount = 1) {
+        return makeGlobalSet(
+          ASYNCIFY_CATCH_COUNTER,
+          makeBinary(SubInt32,
+                     makeGlobalGet(ASYNCIFY_CATCH_COUNTER, Type::i32),
+                     makeConst(int32_t(amount))));
+      };
+    };
+    CountersBuilder builder(*module_);
+    BranchUtils::BranchTargets branchTargets(func->body);
+
+    // with this walker we will assign count of enclosing catch block to
+    // each expression
+    // ... - 0
+    // catch
+    //   ... - 1
+    //   catch
+    //     ... - 2
+    std::unordered_map<Expression*, int> expressionCatchCount;
+    struct NestedLevelWalker
+      : public PostWalker<NestedLevelWalker,
+                          UnifiedExpressionVisitor<NestedLevelWalker>> {
+      std::unordered_map<Expression*, int>* expressionCatchCount;
+      int catchCount = 0;
+
+      static void doStartCatch(NestedLevelWalker* self, Expression** currp) {
+        self->catchCount++;
+      }
+
+      static void doEndCatch(NestedLevelWalker* self, Expression** currp) {
+        self->catchCount--;
+      }
+
+      static void scan(NestedLevelWalker* self, Expression** currp) {
+        auto curr = *currp;
+        if (curr->_id == Expression::Id::TryId) {
+          self->expressionCatchCount->insert(
+            std::make_pair<>(curr, self->catchCount));
+          auto& catchBodies = curr->cast<Try>()->catchBodies;
+          for (Index i = 0; i < catchBodies.size(); i++) {
+            self->expressionCatchCount->insert(
+              std::make_pair<>(catchBodies[i], self->catchCount));
+            self->pushTask(doEndCatch, currp);
+            self->pushTask(NestedLevelWalker::scan, &catchBodies[i]);
+            self->pushTask(doStartCatch, currp);
+          }
+          self->pushTask(NestedLevelWalker::scan, &curr->cast<Try>()->body);
+          return;
+        }
+
+        PostWalker<NestedLevelWalker,
+                   UnifiedExpressionVisitor<NestedLevelWalker>>::scan(self,
+                                                                      currp);
+      }
+
+      void visitExpression(Expression* curr) {
+        expressionCatchCount->insert(std::make_pair<>(curr, catchCount));
+      }
+    };
+    NestedLevelWalker nestedLevelWalker;
+    nestedLevelWalker.expressionCatchCount = &expressionCatchCount;
+    nestedLevelWalker.walk(func->body);
+
+    // with this walker we will handle those changes of counter:
+    // - entering into catch (= pop)  +1
+    // - return                       -1
+    // - break                        -1
+    // - exiting from catch           -1
```

Comment:
Maybe I don't understand what the counter is. In the previous class we counted the nesting level of catch bodies. Why do we want to count say a break as -1? (and why -1 and not some other value?)

Reply:
We need to understand are we inside a catch block when we call 'unwind'. No matter are we inside only top-level catch or some nested catch. Obviously entering or exiting a nested catch didn’t change the fact that we were already in the catch block. Because of that I change counter only when entering/exiting inside top-level catches. 

The comment is not clear, I think it should be:
```
    // - entering into catch (= pop)                +1
    // - return                                     -1
    // - break (target outside of top-level catch)  -1
    // - break (target inside of top-level catch)    0 (ignored)
    // - exiting from catch                         -1
```

Reply:
Even more:

```
    // - entering top-level catch (= pop)           +1
    // - entering nested catch (= pop)               0 (ignored)
    //
    // - return inside top-level/nested catch       -1
    // - return outside top-level/nested catch       0 (ignored)
    //
    // - break target outside of top-level catch    -1
    // - break target inside of top-level catch      0 (ignored)
    // - break outside top-level/nested catch        0 (ignored)
    //
    // - exiting from top-level catch               -1
    // - exiting from nested catch                   0 (ignored)
```

Reply:
Thanks, I see. I think what's confusing me is that more than one class is used to achieve this? I think there's a simpler way. See `ExpressionStackWalker` - you can use that traversal to get the stack of all the parents. Then you can simply walk up the stack to see if you are inside a catch body or not.

Reply:
Done!